### DISCUSSION TITLE
Is it possible to register service without using extension of Container?

### DISCUSSION THREAD

**Discussion by lukakilas** [hmlongco/Factory#193] [Show and tell]
In my new swiftUI project I'm using modular approach. Each module is framework type. ModuleA needs to receive baseURL from the App.

```
// Inisde The Main App Module
extension Container {
    var moduleA: Factory<ModuleA> {
        self { ModuleAImpl(baseURL: URL(string: "")!) }
    }
}

// Inside ModuleA
public protocol ModuleA {
    func getSomeSpecificView()
}

public struct ModuleAImpl: ModuleA {
    
    public var baseURL: URL
    
    public init(baseURL: URL) {
        self.baseURL = baseURL
        // I'd like to have possibility of registering moduleAClient from here, where I have the base URL received from parent app.
    }
    
    public func getSomeSpecificView() {
        
    }
}

class ModuleAClient: APIClient {
    init(baseURL: URL) {
        // do some stuff
    }
}


extension Container {
    // The problem is that I don't have url here.
    var moduleAClient: Factory<APIClient> {
        self { ModuleAClient(baseURL: <#T##URL#>) }
    }
}
```

My question is, if there is a possibility of registering service with other way, except declaring it in extension of Container?
or how should I handle this particular case?

**Comment by MrSkwiggs** [comment]
Yes, use a `ParameterFactory`:

```swift
extension Container {
    var moduleAClient: ParameterFactory<URL, APIClient> {
       self { ModuleAClient(baseURL: $0) }
    }
}
```

Then, to resolve it:

```swift
let baseURL = <get your base URL>
let client = Container.shared.moduleAClient(baseURL)
```

**Reply by lukakilas** [reply]
Thank you very much! This solution is helpful for me.
As I guess there is no possibility to register services without declaring them inside extension, is that right?

**Reply by MrSkwiggs** [reply]
The `Container` type is only there to provide a convenient default container, on which you can easily add extensions.
But you could also declare your own container like explained [here](https://hmlongco.github.io/Factory/documentation/factory/containers). Then you would declare dependencies directly into your own Container instance
